dali/imgcodec/tools/imagemagick_test.cc

```diff
@@ -0,0 +1,283 @@
+                                   std::filesystem::recursive_directory_iterator &it) {
+  std::vector<std::string> filenames;
+  filenames.reserve(env.batch);
+  while (it != std::filesystem::end(it) && filenames.size() < env.batch) {
+    const auto& entry = *(it++);
+    if (entry.is_regular_file()) {
+      const auto path = entry.path().string();
+      if (std::regex_match(path, env.filter))
+        filenames.push_back(path);
+    }
+  }
+  return filenames;
+}
+
+std::optional<TensorShape<>> scan_imagemagick_shape(FILE *pipe) {
+  int w, h, c;
+  char tmp[16];
+  std::string colors;
+  if (fscanf(pipe, "%d %d %16s", &w, &h, tmp) != 3) return {};
+  colors = tmp;
+
+  if (colors == "srgb" || colors == "rgb") {
+    c = 3;
+  } else if (colors == "srgba" || colors == "rgba" || colors == "cmyk") {
+    c = 4;
+  } else if (colors == "gray") {
+    c = 1;
+  } else {
+    return {};
+  }
+
+  TensorShape<> imagemagick_shape = {h, w, c};
+  return imagemagick_shape;
+}
+
+void process(Env &env, std::vector<std::string> filenames) {
+  std::ostringstream log, out;
+
+  std::ostringstream cmd;
+  cmd << env.identify_path << " -quiet -format  \"%w %h %[channels]\\n\"";
+  for (const std::string &f : filenames) {
+    cmd << " " << f;
+  }
+  FILE* pipe;
+  {
+    const std::lock_guard lock(env.mutex);
+    pipe = popen(cmd.str().c_str(), "r");
+  }
+
+  for (const std::string &filename : filenames) {
+    auto imagemagick_shape = scan_imagemagick_shape(pipe);
+    if (!imagemagick_shape) {
+      fail(log, filename, "Unable to parse ImageMagick's output");
+      continue;
+    }
+
+    if (env.print) {
+      out << filename << "\t" << *imagemagick_shape << std::endl;
+      continue;
+    } 
+    try {
+      auto imgcodec_shape = env.imgcodec_tester.shape_of(filename);
+
+      if (!imgcodec_shape) {
+        fail(log, filename, "Imgcodec failed to parse");
+        continue;
+      }
+      
+      if (*imagemagick_shape == *imgcodec_shape) {
+        ok(log, filename);
+      } else {
+        std::ostringstream ss;
+        ss << "Expected " << *imagemagick_shape << " but got " << *imgcodec_shape;
+        fail(log, filename, ss.str());
+      }
+    }
+    catch (const std::exception &e) {
+      std::ostringstream ss;
+      ss << "Imgcodec raised an exception: " << e.what();
+      fail(log, filename, ss.str());
+    }
+  }
+
+  {
+    const std::lock_guard lock(env.mutex);
+    std::cerr << log.str();
+    std::cout << out.str();
+  }
+}
+
+void run(Env &env) {
+  if (!std::filesystem::is_directory(env.directory)) {
+    process(env, {env.directory});
+    return;
+  }
+
+  auto directory_it = std::filesystem::recursive_directory_iterator(env.directory);
+  ThreadPool pool(env.jobs, -1, false, "imagemagick");
+
+  while (directory_it != std::filesystem::end(directory_it)) {
+    auto batch = get_batch(env, directory_it);
+    pool.AddWork([=, &env](int tid){
+      process(env, batch);
```

Comment:
This will cause race on env.identify_path in process.

Reply:
I'm not sure I see the race. `env.identify_path` is only read and not written to. Are parallel reads a problem here?